File: 022_generate_parentheses/Solution.java
import java.util.List;
import java.util.ArrayList;

public class Solution {
    List<String> result = new ArrayList<>();

    public List<String> generateParenthesis(int n) {
       dfs(n, n, "");

       return result;
    }

    public void dfs(int l, int r, String curStr) {
        if (l == 0 && r == 0) {
            result.add(curStr);
            return;
        }

        if (l > 0) dfs(l - 1, r, curStr + "(");
        if (r > l) dfs(l, r - 1, curStr + ")");
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        List<String> res = solution.generateParenthesis(3);
        for (String str : res) {
            System.out.println(str);
        }
    }
}

File: 022_generate_parentheses/Solution.py
class Solution:
    def __init__(self):
        self.result = []

    def generateParenthesis(self, n: int) -> list:
        self.dfs(n, n, "")
        return self.result

    def dfs(self, l: int, r: int, cur_str: str) -> None:
        if (l == 0 and r == 0):
            self.result.append(cur_str)
            return

        if (l > 0): self.dfs(l - 1, r, cur_str + '(')
        if (r > l): self.dfs(l, r - 1, cur_str + ')')


if __name__ == '__main__':
    solution = Solution()
    result = solution.generateParenthesis(3)
    for string in result:
        print(string)

